### PR TITLE
Add user max sessions options

### PR DESCRIPTION
**Pull request by joshuaboniface**
**Changes**

Adds the ability to set a maximum number of active sessions per account, and limit logins based on this count.

**Issues**

Implements Fider #59

### CODE REVIEW

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 1]
```diff
@@ -1484,6 +1484,14 @@ private async Task<AuthenticationResult> AuthenticateNewSessionInternal(Authenti
                 throw new SecurityException("User is not allowed access from this device.");
             }
 
+            int sessionsCount = Sessions.Where(i => string.Equals(i.UserId, user.Id)).ToList().Count;
```
Comment (crobibero) [inline Emby.Server.Implementations/Session/SessionManager.cs:1487]:
```suggestion
            int sessionsCount = Sessions.Count(i => string.Equals(i.UserId, user.Id, StringComparison.OrdinalIgnoreCase));
```

Comment (joshuaboniface) [reply Emby.Server.Implementations/Session/SessionManager.cs:1487]: These are GUIDs so they *should* have the same case, but no harm.

Comment (joshuaboniface) [reply Emby.Server.Implementations/Session/SessionManager.cs:1487]: This change actually doesn't work for that exact reason:

```
Session/SessionManager.cs(1487,67): error CS1503: Argument 1: cannot convert from 'System.Guid' to 'string' [/home/vsts/work/1/s/Emby.Server.Implementations/Emby.Server.Implementations.csproj]
```

Reverted the suggestion and leaving as-is.

Comment (cvium) [reply Emby.Server.Implementations/Session/SessionManager.cs:1487]: 1. They are both Guids. No need to use `string.Equals`. Could just do `i.UserId.Equals(user.Id)`.
2. `ToList()` is a waste of an allocation. Just call `Count()` directly.

Comment (joshuaboniface) [reply Emby.Server.Implementations/Session/SessionManager.cs]: Confirmed this simpler format works great :+1:

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 2]
```diff
@@ -1484,6 +1484,14 @@ private async Task<AuthenticationResult> AuthenticateNewSessionInternal(Authenti
                 throw new SecurityException("User is not allowed access from this device.");
             }
 
+            int sessionsCount = Sessions.Count(i => i.UserId.Equals(user.Id));
+            int maxActiveSessions = user.MaxActiveSessions;
+            _logger.LogInformation("Current/Max sessions for user {User}: {Sessions}/{Max}", user.Username, sessionsCount, maxActiveSessions);
```
Comment (crobibero) [inline Emby.Server.Implementations/Session/SessionManager.cs:1489]: Probably only log this if the max sessions is set

Comment (joshuaboniface) [reply Emby.Server.Implementations/Session/SessionManager.cs:1489]: I debated this, but I think the information is important and useful regardless. I've long thought we log far too much useful data like this.

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Model/Users/UserPolicy.cs [hunk 3]
```diff
@@ -144,6 +146,8 @@ public UserPolicy()
 
             LoginAttemptsBeforeLockout = -1;
 
+            MaxActiveSessions = -1;
```
Comment (barronpm) [inline MediaBrowser.Model/Users/UserPolicy.cs:149]: Should be 0?

File: Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿#pragma warning disable CS1591
+#pragma warning disable SA1601
+
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace Jellyfin.Server.Implementations.Migrations
+{
+    public partial class MaxActiveSessions : Migration
```
Comment (barronpm) [inline Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs:8]: Should be `AddMaxActiveSessions` to be in line with the naming of other migrations

**Review by barronpm — CHANGES_REQUESTED** [review]
Looks good, pending discussion on whether we should make MaxActiveSessions nullable

File: MediaBrowser.Model/Users/UserPolicy.cs (hunk 3 above)
Comment (joshuaboniface) [reply MediaBrowser.Model/Users/UserPolicy.cs:149]: OK fine lol

**Comment by joshuaboniface** [comment]
>pending discussion on whether we should make MaxActiveSessions nullable

As mentioned in chat, while I'm not adverse to this, it also resulted in some weirdness when I tried this earlier in the process. If this is the only one that should be nullable, we can change it, but I'd need advice on how to do so on the frontend especially.

File: Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs (hunk 4 above)
Comment (joshuaboniface) [reply Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs:8]:
```
-rw-r--r-- 1 joshua operator 1.7K Sep 24 21:15 20200905220533_FixDisplayPreferencesIndex.cs
-rw-r--r-- 1 joshua operator  16K Sep 24 21:15 20200905220533_FixDisplayPreferencesIndex.Designer.cs
```

This one does not use the format and was the one I based it on. I don't mind regenerating though.

Comment (joshuaboniface) [reply Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs:8]: Reading comprehension fail again, "Add" vs "Fix" makes sense.

Comment (joshuaboniface) [reply Jellyfin.Server.Implementations/Migrations/20201004171403_MaxActiveSessions.cs]: Renamed and verified that this matches the expected values.

**Review by barronpm — APPROVED** [review]

**Pull request merged**